src/test/java/echoserver/InputValidatorTest.java

```diff
@@ -0,0 +1,46 @@
+package echoserver;
+
+import org.junit.Test;
+import static org.junit.Assert.*;
+
+public class InputValidatorTest {
+    @Test
+    public void returnsTrueIfInputEqualsQuit() {
+        assertTrue(InputValidator.isQuit("quit"));
+    }
+
+    @Test
+    public void returnsFalseIfInputIsNotQuit() {
+        assertFalse(InputValidator.isQuit("hello"));
+    }
+
+    @Test
+    public void returnsTrueRegardlessOfCaseSensitivityForQuit() {
+        assertTrue(InputValidator.isQuit("QuIT"));
+    }
+
+    @Test
+    public void returnsTrueIfQuitHasSpacesAtTheEnd() {
+        assertTrue(InputValidator.isQuit("quit    "));
+    }
+
+    @Test
+    public void returnsTrueIfQuitHasNewLineAtTheEnd() {
+        assertTrue(InputValidator.isQuit("quit\n"));
```

Comment:
I'm not opposed to having this test, but did it cause a change in the production code when this test was added?  It's pretty natural to call `String#trim` whenever you need to trim whitespace of any kind, and that would already handle this newline character or the spaces in the test above.

Just curious.

Reply:
When we were TDDing this class, we wrote the `returnsTrueIfQuitHasSpacesAtTheEnd()` test to verify the `.trim()` functionality, and then wrote `returnsTrueIfQuitHasNewLineAtTheEnd()` while taking production use cases into account. We definitely did not need to write the first test, as it did not cause a change in the production code!

Reply:
That's cool.  I'm not opposed to keeping the test if you want to make sure it covers both cases now and in the future.  Sometimes I do that; I try not to be too dogmatic about it.

It's also perfectly fine to write a test, find that it doesn't cause the production code to break, and remove the test (perhaps generalizing the description in the previous test). In this case, one possible progression is to start with "it removes spaces at the end" and then change it to "it removes all whitespace" when you find you can't break the code after adding `String#trim`.

This is really just an opportunity to discuss and reflect, possibly seeing some of the nuances and grey areas when test-driving code.  No action required.